crypto/init.c

```diff
@@ -729,3 +683,52 @@ int OPENSSL_atexit(void (*handler)(void))
     return 1;
 }
 
+/*
+ * Library  destructor (a.k.a. _fini()). Code here is derived
+ * from stuff found in providers/fips/self_test.c
+ *
+ * We do kind of  best effort cleanup. If given platform/compiler
+ * does not support module destructors then its up to application
+ * to explicitly call OPENSSL_cleanup();
+ */
+#if defined(__GNUC__) && !defined(_AIX)
+
+void ossl_call_cleanup(void);
+
+void __attribute__((destructor))
+ossl_call_cleanup(void)
+{
+    if (ossl_no_cleanup == 0)
+        OPENSSL_cleanup();
+}
+
+#elif (defined(__sun) || defined(_AIX)) && !defined(__GNUC__)
+void ossl_call_cleanup(void);
+# pragma fini(ossl_call_cleanup)
+void ossl_call_cleanup(void)
+{
+    if (ossl_no_cleanup == 0)
+        OPENSSL_cleanup();
+}
+
+#elif defined(__hpux)
+# pragma fini "ossl_call_cleanup"
+void ossl_call_cleanup(void)
+{
+    if (ossl_no_cleanup == 0)
+        OPENSSL_cleanup();
+}
+#elif defined(__TANDEM)
+/* Method automatically called by the NonStop OS prior to unloading the DLL */
+void __TERM__cleanup(void)
+{
+    if (ossl_no_cleanup == 0)
+        OPENSSL_cleanup();
+}
+#else
+void _fini(void)
```

Comment:
This is different to the providers/fips/self_test.c version. In the fips version the function does not get defined at all - but here you seem to define it anyway and call it `_fini`. Why is that?

Reply:
This is the fallback for all other platforms/linkers which are not covered by `if defined()` cases here. I did test this variant with `clang -std C89` using a simple unit test and it did the right thing.